### ISSUE TITLE
[BUG] Storage is allocating more than overcommit-config

### ISSUE THREAD

**Issue by midhunthampi** [harvester/harvester#3083]
**Describe the bug**

**Expected behavior**
We need to keep Storage scheduling below than max. storage available when we keep overcommit config is 100%

**Environment**
 - Harvester ISO version: 1.1.0
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge C6420):

Depends on: https://github.com/longhorn/longhorn/issues/5149

**Comment by ibrokethecloud** [comment]
are you please able to provide more details of the operation you are trying to perform and what should have been the expected result and what is actually happening?

Any additional information like a support bundle which we can analyze to check the system state would be appreciated.

**Comment by midhunthampi** [comment]
<img width="1201" alt="Screen Shot 2022-11-01 at 4 32 31 PM" src="https://user-images.githubusercontent.com/62717435/199362951-5a7f70c3-3e50-4b11-8f84-eb3cadbcdc77.png">

This is the image, which we updated in settings to use Storage 100%

**Comment by midhunthampi** [comment]
But still, harvester assign the storage more than disk size (100%). It's creating issue and making all VM's in that datastore going down when storage become full.

<img width="967" alt="Screen Shot 2022-11-01 at 4 30 03 PM" src="https://user-images.githubusercontent.com/62717435/199363192-13468a6b-a12f-4694-ad41-1ad8aa4cee84.png">

Please see the Red highlighted storage

**Comment by midhunthampi** [comment]
At above screenshot, Our storage size is 3.5 TB but its scheduled till 4.5TB even after we set Storage overcommit to 100%

**Comment by TheiLLeniumStudios** [comment]
@midhunthampi I think a 100% overcommit means that it can provision a total of 7TB. You should change the overcommit for storage to 0 to disable overcommit entirely

**Comment by midhunthampi** [comment]
admission webhook "validator.harvesterhci.io" denied the request: Cannot undercommit. Should be greater than or equal to 100 but got 0

Its not allowing to schedule any value below than 100%.

**Comment by Martin-Weiss** [comment]
Overcommit 100% means you can provision 100% of what is available. 
Less than 100% allows less to be provisioned. 
0% would mean to disable storage ;-). 

Depending on snapshots even with 100% set more can and will be used depending on snapshots..
See also https://longhorn.io/docs/1.3.1/volumes-and-nodes/volume-size/ 

So in case you do not want to use more than 100% - you should set the overcommit below 100% based on the % you want to reserve for snapshots.

**Comment by midhunthampi** [comment]
But its not allowing any percentage below than 100.

admission webhook "validator.harvesterhci.io" denied the request: Cannot undercommit. Should be greater than or equal to 100 but got 80

**Comment by Martin-Weiss** [comment]
Thanks for testing - in this case I assume this is a bug in the webhook. Can you try to set 80% with the longhorn UI?

**Comment by midhunthampi** [comment]
Thanks for testing - in this case I assume this is a bug in the webhook. Can you try to set 80% with the longhorn UI? - yes, I tried to set 80%.

**Comment by Vicente-Cheng** [comment]
Hi @midhunthampi,
Did you remember when did you adjust the overcommit ratio of storage?
The default overcommit ratio is 200. That means the storage could schedule more than 100%.
(I would like to confirm that when the storage allocation is scheduled, is an overcommit ratio 200% or 100%?)

For the webhook setting of overcommit, I thought that is blocked when you want to any percentage below 100 on the harvester UI.
But You could change the percentage below 100 on Longhorn UI.

**Comment by midhunthampi** [comment]
@Vicente-Cheng 

Did you remember when did you adjust the overcommit ratio of storage? We have adjusted the Storage overcommit ratio after the cluster setup itself

The default overcommit ratio is 200. That means the storage could schedule more than 100%. - When we setup the cluster, it was 200%. which means if we have 3TB disk, it allowed till around 6TB or more - In these case, Once disk get full,all VM's in these datastore going down/affecting. Thats why we decided to keep 100%.

But even after keep 100%, it's allowing some percentage more than 100%.

Now we have set Storage overcommit ratio to 80% in LongHorn. Lets see how it will work.

Basically, we are looking for solution, where datastore never get full and VM's will be running all time w/o any impact

**Comment by Vicente-Cheng** [comment]
> @Vicente-Cheng
> 
> Did you remember when did you adjust the overcommit ratio of storage? We have adjusted the Storage overcommit ratio after the cluster setup itself
> 
> The default overcommit ratio is 200. That means the storage could schedule more than 100%. - When we setup the cluster, it was 200%. which means if we have 3TB disk, it allowed till around 6TB or more - In these case, Once disk get full,all VM's in these datastore going down/affecting. Thats why we decided to keep 100%.
> 
> But even after keep 100%, it's allowing some percentage more than 100%.
> 
> Now we have set Storage overcommit ratio to 80% in LongHorn. Lets see how it will work.
> 
> Basically, we are looking for solution, where datastore never get full and VM's will be running all time w/o any impact

Hi @midhunthampi 
OK, we could see how it works when you set up the overcommit ratio to 80%. Just update what you see. That would be helpful!

IIRC, longhorn would reserve some space to make sure they could work well. So I thought that even if you set the overcommit
to 200% the disk should be not filled full. Or do you meet this problem (disk full with overcommit) in real cases?

Thanks!

**Comment by Martin-Weiss** [comment]
@Vicente-Cheng  - not sure how you calculate the „200% should not fill up the disk“.. Longhorn reserves 30% per node.. so with 200% the disks will fill up..

3 nodes with 100 GB Longhorn disk = 3 * 70 GB available. So with 100% we could create 7 * 10 GB Volumes with Replica=3.
Creating with 200% means 140 GB * 3.. as long as there is no content in the volumes (thin provisioned) - all is fine but once the volumes fill up - the Longhorn disks will run out of available disk space.
„Over-provisioning“ is a risk and needs proper monitoring and fast pro-active reaction (buy new hard disks or cluster nodes to expand before running out of disk space)… so 80% is a more safe approach.

And keep in mind that the over-provisioning limit (80/100/200%) is only looked at during provisioning. Snapshots can and  will consume additional space (that and the OS is why there is the 30% reservation AFAIK) - and in this case the space will fill up, too…

Please correct my if my calculation and assumptions are not correct ;-). Thanks!

**Comment by Vicente-Cheng** [comment]
Hi @Martin-Weiss,
Yap, you are right. I got some misunderstanding.
Thanks for your example, that is true. Disks would fill up in a scenario as you mention above.
And I also do some tests in my own environment, which can fill up the disk.

@midhunthampi, as Martin says.
The 200% overcommit would have chances to fill up your disks. You may just set the lower overcommit ratio to reduce this situation. But if you have many snapshots, the reserved disk space would also be filled up.
The Longhorn snapshot is CoW. That means in some specific scenarios the reserved disk space would be not enough.

If you want your datastore never to get full, maybe try to reduce the overcommit ratio and also reduce the number of snapshots. That would be safer, but never 100% safe.

**Comment by midhunthampi** [comment]
Hi @Martin-Weiss  @Vicente-Cheng 

We have set storage overcommit ratio is 80% in Long Horn and monitoring now. How we can reduce the snapshot numbers. We didn't configure snapshot for any volume yet. If its automatically configure, how we can adjust it ?

**Comment by Martin-Weiss** [comment]
> Hi @Martin-Weiss @Vicente-Cheng
> 
> We have set storage overcommit ratio is 80% in Long Horn and monitoring now. How we can reduce the snapshot numbers. We didn't configure snapshot for any volume yet. If its automatically configure, how we can adjust it ?

I do not have all the details in case of Harvester, here - but the standard way of configuring and maintaining snapshots is based on Longhorn - https://longhorn.io/docs/1.3.2/snapshots-and-backups/ and you can see the snapshots in the Longhorn UI on the volumes. It is possible to setup scheduled snapshots / backups in Longhorn and to automatically cleanup previous versions based on these schedules - https://longhorn.io/docs/1.3.2/snapshots-and-backups/scheduling-backups-and-snapshots/ . In Harvester we just got the feature to create snapshots on VMs https://docs.harvesterhci.io/v1.1/vm/backup-restore#vm-snapshot--restore but I did not look into the details on how this interacts with volume snapshots.
I believe one of the most important point regarding storage management in Harvester is to monitor the disk space usage and setup alerting and be pro-active in adding more capacity when needed.

**Comment by Vicente-Cheng** [comment]
> Hi @Martin-Weiss @Vicente-Cheng
> 
> We have set storage overcommit ratio is 80% in Long Horn and monitoring now. How we can reduce the snapshot numbers. We didn't configure snapshot for any volume yet. If its automatically configure, how we can adjust it ?

Hi @midhunthampi, 
Thanks for the detailed information from @Martin-Weiss.

Most snapshot operations of the user are described with his reply.
If you do not need to snapshot your data, maybe keeping the overcommit ratio at 80% would be fine.

Something you need to know is that longhorn would take some snapshots when replica rebuilding, but these snapshots would be cleaned up when the rebuild is complete. Just a note, don't worry about it.

If you want to automatically snapshot your data, in my opinion do not keep too many snapshots for one volume (maybe 3~5 snapshots are enough). That would keep the performance and save your capability.

**Comment by Martin-Weiss** [comment]
@Vicente-Cheng  - I agree that we consume less capacity and have better performance the less snapshots we have but in the end the requirements will define how many snapshots should be created and kept. (RTO / RPO). If the recovery point objective is "need to be sure to be sure that we can restore to the point we had an hour ago - then we need to create snapshots every hour". And if the reaction time for data-loss or rollback requirements is 24h we need to keep the snapshots for 24h - so we might end up with 25 snapshots.. And for the capacity / reserved space we need to know how much data changes per hour so we can sum up how much capacity we need reserved.. 

Thoughts about examples:

- Volume 1 GB and has 1 GB change every hour  -> ~25 GB required even though the volume has 1 GB, only..
- Volume has 10 GB and 1 GB change every hour -> ~35 GB required (ingnoring the thin provisioning - but might not count as long as we do not support trim)
- Volume has 10 GB and 1 MB change every hour -> ~11 GB required

**Comment by midhunthampi** [comment]
@Vicente-Cheng 

nd if the reaction time for data-loss or rollback requirements is 24h we need to keep the snapshots for 24h - so we might end up with 25 snapshots. - Is anyway how we can change the frequency of snapshot creation or reduce the overall retention time ?

**Comment by Martin-Weiss** [comment]
@midhunthampi  - I believe this is documented in the Longhorn documentation - https://longhorn.io/docs/1.3.2/snapshots-and-backups/scheduling-backups-and-snapshots/. 
AFAIK we do not have recuring jobs within Harvester per default but I assume you can create them based on your needs.

**Comment by midhunthampi** [comment]
@Martin-Weiss @Vicente-Cheng 

Even after we set Storage Over Provisioning 80% in LongHorn, It's allowing to allocate storage more than 100%. Please see below screen shot.

It's in Red state but allocation is more than max. storage.

<img width="951" alt="Screen Shot 2022-11-29 at 4 49 37 PM" src="https://user-images.githubusercontent.com/62717435/204681018-5ccb6c74-d339-4d4b-83ed-298b9e2d099f.png">

**Comment by midhunthampi** [comment]
![Uploading Screen Shot 2022-11-29 at 4.57.21 PM.png…]()

**Comment by midhunthampi** [comment]
@Martin-Weiss @Vicente-Cheng 

We are looking for solution like that Storage Scheduled always less than Storage Max

Storage Scheduled < Storage Max.

**Comment by midhunthampi** [comment]
Also why its adding all replica's in single disk instead of multiple disks. These hosts having same size other disks as well but its keep adding volume in single disk.

**Comment by Martin-Weiss** [comment]
> We are looking for solution like that Storage Scheduled always less than Storage Max
> 
> Storage Scheduled < Storage Max.

I believe this would only be possible if snapshots would never be created.

> Also why its adding all replica's in single disk instead of multiple disks. These hosts having same size other disks as well but its keep adding volume in single disk.

Not sure what you mean by that - but Longhorn uses a disk that has sufficient space and tries to distribute the data among different nodes and different disks..

**Comment by midhunthampi** [comment]
@Martin-Weiss 

Not sure what you mean by that - but Longhorn uses a disk that has sufficient space and tries to distribute the data among different nodes and different disks..--->>>>>>>>

Longhorn should distribute the data among different nodes and disks but its not happening during VM creation 

Ex: -
Above case in screenshot, where storage scheduled till 9.9TB where max. storage is 7TB even we kept storage allocation percentage is 80% in LongHorn.

In that host, we have 5 x 7TB disks and when we create VM, it was filling one of the above mentioned disk even other disks very less utilized (below 10%).

Therefore, we need to delete replica of volume's from higher utilized disk via longhorn and we able to see Longhorn is creating new replica  and distributing among the different nodes and disks.

So i believe, when we create any volume/VM via harvester, storage overcommit ratio is not working correctly but it seems working via Longhorn when we create new replica's there.

**Comment by Martin-Weiss** [comment]
@midhunthampi  - I do not know the algorithms for distributing data in longhorn - but there is a difference between allocated and requested and maybe this also affects how the data is distruibuted.

There is this document describing the thin provision challenge https://longhorn.io/docs/1.3.2/volumes-and-nodes/volume-size/ and there is the feature to re-distribute data https://longhorn.io/docs/1.3.2/references/settings/#replica-auto-balance . 
Temporarily removing nodes (due to maintenance) also re-distributes data so the distribution is not 100% balanced..

**Comment by Vicente-Cheng** [comment]
Hi @midhunthampi, 
I saw your picture. Could you check from Longhorn UI for snapshots? (Maybe you could screenshot the snapshot column)

Then, could you try to ls -lR and du for the oversize directory?

And last, could you also provide the support bundle for us to investigate this problem?
Thanks!

**Comment by midhunthampi** [comment]
@Vicente-Cheng  @Martin-Weiss 

Let me give overall picture of the issue.
1. This is 4 node new cluster - NO VM's added yet
2. We are creating VM via terraform (around 40TB size) VM
3. We able to see most of these VM volume going to same disk in host even other disks in the host having space
4. We kept 2 replica's for all volume and see both replica's are creating in different hosts.No issue with replica creation in different hosts
5. Only issue Harvester creating one of replica of all these volume's are adding in same disk of Host.
6. We kept the storage overcommit ratio is 100% in Harvester. Ever after it allocating volume more than disk size.
7. We kept the storage overcommit percentage in LongHorn is 80% and its working fine if we create any volume from Longhorn.
8. Also we see that, if we delete one of replica from over utilized disk via longhorn, it will add the replica in other disks perfectly. 
9. So now whenever we create VM, we need to go there in Longhorn and delete replica's manually to come these volume's correctly

**Comment by midhunthampi** [comment]
<img width="962" alt="Screen Shot 2022-12-01 at 1 56 51 PM" src="https://user-images.githubusercontent.com/62717435/205350052-1a3be39d-9e18-4434-b9fd-5b3c776f4e4f.png">

Here is the screenshot of fresh first VM, we created in these cluster. NO data and snapshots with these Volumes. But harvester adding the volume's in same disk

**Comment by Martin-Weiss** [comment]
What do you mean with "2 replica" - per default Longhorn/Harvester creates 3 replicas.. in case you want to have 2 replicas, only per default and always - you need to create a new storage class with replica=2 and use that one when uploading images and when creating new VMs.

Why do you have multiple disks in the hosts instead of creating a single Raid device on the hardware side of things?

Regarding the schedule that all / most is scheduled on the first disk, only - just guessing - but this might also be a race condition when you create many volumes at the same point in time and they are scheduled "in parallel". In this case the first disk looks "empty" for all of them..
-> Do you see this also when you create one VM after the other and when leaving 5-10 Minutes between each of them?

**Comment by midhunthampi** [comment]
What do you mean with "2 replica" - per default Longhorn/Harvester creates 3 replicas.. in case you want to have 2 replicas, only per default and always - you need to create a new storage class with replica=2 and use that one when uploading images and when creating new VMs. - We already have new storage class with 2 replica's and it's working also.

Why do you have multiple disks in the hosts instead of creating a single Raid device on the hardware side of things? - Already we are keeping replica. On top of that, if we are using Hardware Raid, it's loss of all disk space. And what about performance if we RAID on the device ?

Regarding the schedule that all / most is scheduled on the first disk, only - just guessing - but this might also be a race condition when you create many volumes at the same point in time and they are scheduled "in parallel". In this case the first disk looks "empty" for all of them.. -- Ok, One thing , we are creating the VM via terraform. We mainly see issue when create VM via terraform.

Do you see this also when you create one VM after the other and when leaving 5-10 Minutes between each of them? We see same issue. But we see that Longhorn is working correctly if we delete any replica from there. It creating/ distributing the data..

**Comment by Martin-Weiss** [comment]
Ok - so the main issue really seems to be that Longhorn does not distribute the „new volumes“ automatically when creating VMs.  And just to be sure to be sure - do I get it right that you see this issue also in case you create a new VM manually via the UI?
The problem might also be related to the image used. Maybe there is a initial clone / linked clone of the OS image that will go to the same disk.. but on this level I am not deep enough in the creation process of a VM and we need someone from engineering to dig deeper.

**Comment by Vicente-Cheng** [comment]
> @Vicente-Cheng @Martin-Weiss
> 
> Let me give overall picture of the issue.
> 
> 1. This is 4 node new cluster - NO VM's added yet
> 2. We are creating VM via terraform (around 40TB size) VM
> 3. We able to see most of these VM volume going to same disk in host even other disks in the host having space
> 4. We kept 2 replica's for all volume and see both replica's are creating in different hosts.No issue with replica creation in different hosts
> 5. Only issue Harvester creating one of replica of all these volume's are adding in same disk of Host.
> 6. We kept the storage overcommit ratio is 100% in Harvester. Ever after it allocating volume more than disk size.
> 7. We kept the storage overcommit percentage in LongHorn is 80% and its working fine if we create any volume from Longhorn.
> 8. Also we see that, if we delete one of replica from over utilized disk via longhorn, it will add the replica in other disks perfectly.
> 9. So now whenever we create VM, we need to go there in Longhorn and delete replica's manually to come these volume's correctly

Hi @midhunthampi, 
As for your whole flow, could I briefly describe it as follows (if I got something wrong, please correct me):
1. Create 4 node cluster, then create VM from terraform (At this moment, overcommit is still 200%)
2. VM's volume always schedules to the same disk.
3. Reduce the replica to `2`, volumes are still scheduled to the same disk.
4. change the overcommit ratio to 100%, then see the situation with oversized
5. change the overcommit ratio to 80%, working fine with new create volume (only created from Longhorn? if you created on the harvester side, does it work well?)
6. Now, manually delete one volume replica that could help schedule the suitable disk.

If the above flow is correct, I thought the disk oversize was caused by the previous VMs created from terraform.
Longhorn would calculate the disk space (with overcommit ratio) with scheduling replica. So that may your disk is oversized.

Manually deleting the replica would make longhorn reschedule the replica to another disk (because the previous disk is already oversized, longhorn would not schedule the replica to this).

**Comment by Martin-Weiss** [comment]
@Vicente-Cheng  - doesn´t / shouldn´t longhorn distribute new volumes automatically over all available disks in a given node?

**Comment by Vicente-Cheng** [comment]
> @Vicente-Cheng - doesn´t / shouldn´t longhorn distribute new volumes automatically over all available disks in a given node?

Ah, yap. I looked into the code again. It should consider the whole disks on the nodes.

**Comment by Martin-Weiss** [comment]
So then it might be some defect?

**Comment by Vicente-Cheng** [comment]
> defect

I mean that from the code, the longhorn manager would consider whole disks to find the disk with the most available capability to schedule a replica.
I would check it more with code to explain why this situation (using the same disk continuously) is in the @midhunthampi's environment.

**Comment by midhunthampi** [comment]
@Vicente-Cheng 

As for your whole flow, could I briefly describe it as follows (if I got something wrong, please correct me):

    Create 4 node cluster, then create VM from terraform (At this moment, overcommit is still 200%)
    VM's volume always schedules to the same disk.
    Reduce the replica to 2, volumes are still scheduled to the same disk.
    change the overcommit ratio to 100%, then see the situation with oversized
    change the overcommit ratio to 80%, working fine with new create volume (only created from Longhorn? if you created on the harvester side, does it work well?)
    Now, manually delete one volume replica that could help schedule the suitable disk.

There is small correction. Workflow as below

Create 4 node cluster.
Set the Overcommit ration of Storage in Harvester UI is 100% after cluster creation
Set the Longhorn Storage allocation percentage is 80%
Create storage class with 2 replica
Now VM creation is coming. Create the VM withfrom terraform - ex:- 10 x 2 TB disks
We see replica's creating in different node and but its filling the one disk of same node instead of distributing in multiple disks.
Now Manually deleting the replica from Longhorn, its distributing the volume correctly in other disks.

**Comment by midhunthampi** [comment]
One more thing. We able to see same issue with VM creation at UI as well.

**Comment by midhunthampi** [comment]
@Vicente-Cheng  @Martin-Weiss 

I have some questions:-

1. Why it's allowing to allocate more storage even if we set 100% in Harvester UI and 80% in Longhorn UI during VM creation ?
2. Storage overcommit percentage is working perfectly when we create volume from longhorn/ delete replica's from longhorn. Why its not happening during VM creation ?
3. Why auto rebalance of replica is not happening even if we set auto rebalance - best efforts in Longhorn

And have another concerns.

Can we add the feature like auto rebalance the volume's to another disk once one of the disk over provisioned in that node ? It would be greatful and avoid manual deletion of replica from Longhorn everytime.

Otherwise please create this as bug and add the feature to fix this issue. Or please suggest some workaround to fix this issue.

**Comment by midhunthampi** [comment]
this is screenshot when we create VM manually via UI. S. Storage overcommit is not working correctly (Both harvester and Longhorn) during VM creation.

**Comment by midhunthampi** [comment]
![Uploading Screen Shot 2022-12-06 at 1.43.49 PM.png…]()

**Comment by Martin-Weiss** [comment]
@midhunthampi - from my point of view the described behavior (filling up one disk instead of using all disks) is a behavior that does not fit the design nor requirements. So engineering needs to dig into it and find out why it behaves this way and enhance / fix the behavior. -> could you create a support case for this?

**Comment by midhunthampi** [comment]
@Martin-Weiss 

I created this one - https://github.com/harvester/harvester/issues/3256

**Comment by Martin-Weiss** [comment]
> @Martin-Weiss
> 
> I created this one - #3256

Thanks - but my thought was via scc.suse.com 😉

**Comment by midhunthampi** [comment]
@Martin-Weiss Do you want me to create it there ? Just let me know, i can create it

**Comment by Martin-Weiss** [comment]
> @Martin-Weiss Do you want me to create it there ? Just let me know, i can create it

@midhunthampi  - yes - it would be good as this helps for the official process "Support Case" -> "SUSE Jira Ticket" -> "Engineering Prio" ;-)

**Comment by midhunthampi** [comment]
@Martin-Weiss 

Sorry, I am not allowed to raise support case there.

**Comment by Vicente-Cheng** [comment]
Related to https://github.com/longhorn/longhorn/issues/5149, I would work on that.

**Comment by Vicente-Cheng** [comment]
Moving to v1.7.0, we need to recheck this one.

**Comment by Vicente-Cheng** [comment]
cc @harvester/storage, 
We should revisit this issue.
